opt: push limit into offset

ridwanmsharif opened the pull request:
This change pushes the limit into an offset whenever possible.
This shouldn't worsen any plan but does allow the `GetLimitedScans`
rule to fire in more scenarios.

Fixes #30416.
~~This is currently blocked on #38659.~~

Release note: None

Commit:
opt: push limit into offset

This change pushes the limit into an offset whenever possible.
This shouldn't worsen any plan but does allow the GetLimitedScans
rule to fire in more scenarios.

Fixes #30416.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38660)
<!-- Reviewable:end -->

jordanlewis commented:
Thanks. This looks like a bug in distsql physical planning.

Explain plans for your scenario:

```
root@127.0.0.1:51341/defaultdb> explain SELECT * FROM (SELECT * FROM tab ORDER BY a LIMIT 5) ORDER BY a OFFSET 3;
    tree    | field  | description
+-----------+--------+-------------+
  limit     |        |
   │        | offset | 3
   └── scan |        |
            | table  | tab@primary
            | spans  | ALL
            | limit  | 5
(6 rows)

Time: 541µs

root@127.0.0.1:51341/defaultdb> explain(distsql) SELECT * FROM (SELECT * FROM tab ORDER BY a LIMIT 5) ORDER BY a OFFSET 3;
  automatic |                                                                                                                                                                url
+-----------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    true    | https://cockroachdb.github.io/distsqlplan/decode.html#eJyMjzFLBDEUhHt_xTL1E3cRm1S2B-LJaScpcpt3R2A3CXlvQTny3-WSQiyEK-ebZGbeBTF5fnUrC8wnJlhCLmlmkVSuqD_Y-S-YkRBi3vSKLWFOhWEu0KALw-DDHRc-sPNcHkYQPKsLS4vNJayufD-rO4Lwnl0UM9yD8BLWoMPTsD-dhHV4hK2EtOlvi6g7M8xU6fYlB5acovCfEf8lj9US2J-5XytpKzO_lTS3mi737V8DnkW7O3Wxi92qtt79BAAA___H2G4Y
(1 row)

Time: 9.758ms
```

Note that the physical plan combines the offset and limit in a bogus way.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38660)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @ridwanmsharif)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-LjHNq681L5I5p9V9jgN:-LjHNq693qr0Km35bcza:bkx0gbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/147ff262ef992ae2d02d25da5e1ce9a5995aad09/pkg/sql/opt/norm/custom_funcs.go#L1695)):*
> ```Go
> 	firstVal := int64(*first.(*tree.DInt))
> 	secondVal := int64(*second.(*tree.DInt))
> 	return tree.NewDInt(tree.DInt(firstVal + secondVal))
> ```

We should check for overflow here (or somewhere in the rule).


<!-- Sent from Reviewable.io -->

Commit:
opt: check for overflow when applying the PushLimitIntoOffset rule

This change adds an overflow check before applying the rule.

Release note: None

ridwanmsharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38660)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-LjHNq681L5I5p9V9jgN:-LjJkd2x8JFMOdWnIGSE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/147ff262ef992ae2d02d25da5e1ce9a5995aad09/pkg/sql/opt/norm/custom_funcs.go#L1695)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should check for overflow here (or somewhere in the rule).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-:-LjJqs9s4fdzo0SmPvje:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @ridwanmsharif)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1694 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-LjJqXnCBrdYlxpiQocY:-LjJqXnCBrdYlxpiQocZ:bixjoar) ([raw file](https://github.com/cockroachdb/cockroach/blob/a28b31c4282260c3d49dd5d4092672bceda5eb41/pkg/sql/opt/norm/custom_funcs.go#L1694)):*
> ```Go
> // AddConsts adds the numeric constants together. AddConsts assumes the sum
> // will not overflow. Call DoesSumOverflow on the constants to guarantee this.
> func (c *CustomFuncs) AddConsts(first tree.Datum, second tree.Datum) tree.Datum {
> ```

[nit] rename to `AddConstInts`
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1697 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-LjJqeuCEdAgC-O4V74g:-LjJqeuCEdAgC-O4V74h:b-pfvr4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a28b31c4282260c3d49dd5d4092672bceda5eb41/pkg/sql/opt/norm/custom_funcs.go#L1697)):*
> ```Go
> 	firstVal := int64(*first.(*tree.DInt))
> 	secondVal := int64(*second.(*tree.DInt))
> 	sum, _ := arith.AddWithOverflow(firstVal, secondVal)
> ```

check `ok` and `panic(errors.AssertionFailedf(..))` if it's false
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1702 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-LjJqSLyDvWsJ7F3pwlk:-LjJqSLyDvWsJ7F3pwll:bsd754k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a28b31c4282260c3d49dd5d4092672bceda5eb41/pkg/sql/opt/norm/custom_funcs.go#L1702)):*
> ```Go
> 
> // DoesSumOverflow returns true if the addition of the two integers overflows.
> func (c *CustomFuncs) DoesSumOverflow(first tree.Datum, second tree.Datum) bool {
> ```

[nit] I'd invert the result and rename to `CanAddConstInts`
___
*[pkg/sql/opt/norm/rules/limit.opt, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-LjJqFZ2CajqOVBFQ-7-:-LjJqFZ2CajqOVBFQ-70:b-9e09b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a28b31c4282260c3d49dd5d4092672bceda5eb41/pkg/sql/opt/norm/rules/limit.opt#L70)):*
> ```
>     (Const $limit:* & (IsPositiveLimit $limit))
>     $limitOrdering:* & (IsSameOrdering $offsetOrdering $limitOrdering) &
>     ^(DoesSumOverflow $limit $offset)
> ```

[nit] indent this line (4 spaces)
___
*[pkg/sql/opt/norm/testdata/rules/limit, line 479 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-LjJqLXzC19cNztVqvDZ:-LjJqLXzC19cNztVqvD_:bxojxn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a28b31c4282260c3d49dd5d4092672bceda5eb41/pkg/sql/opt/norm/testdata/rules/limit#L479)):*
> ```
> 
> # Using MaxInt64. Do not apply rule when sum overflows.
> opt expect-not=(PushLimitIntoOffset)
> ```

[nit] parens not necessary


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38660#-:-LjMQ9tN0xQePQYCiLdH:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->
